File: src/src_platform_ava/fmg/ava/draw/img/Animator.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.Reactive.Linq;
using Avalonia.Animation;
using fmg.core.img;

namespace fmg.jfx.draw.img {

   public class Animator : IAnimator, IDisposable {

      private class SubscribeInfo {
         public bool active = true;    // enabled?
         public DateTime startTime = DateTime.Now; // start time of subscribe
         public Action<TimeSpan /* time from the beginning of the subscription */> callback;
      }
      private readonly IDisposable _unsubscribe;
      private readonly IDictionary<object /* subscriber */, SubscribeInfo> _subscribers;

      private static Animator _singleton;
      public static Animator getSingleton() { // not synchronized. since should work only in the thread of the UI.
         if (_singleton == null)
            _singleton = new Animator();
         return _singleton;
      }

      private Animator() {
         _subscribers = new Dictionary<object, SubscribeInfo>();
         _unsubscribe = Animate.Timer.Subscribe(
               value => {
                  var currentTime = DateTime.Now;
                  foreach (var item in _subscribers) {
                     var subscribeInfo = item.Value;
                     if (subscribeInfo.active)
                        subscribeInfo.callback(currentTime - subscribeInfo.startTime);
                  }
               },
               ex => { System.Diagnostics.Debug.Assert(false, ex.ToString()); },
               () => { System.Diagnostics.Debug.WriteLine("Closed"); }
            );
      }

      public void Subscribe(object subscriber, Action<TimeSpan /* time from start subscribe */> subscriberCallbackMethod) {
         SubscribeInfo info = _subscribers[subscriber];
         if (info == null) {
            info = new SubscribeInfo();
            info.callback = subscriberCallbackMethod;
            _subscribers.Add(subscriber, info);
         } else {
            info.active = true;
            info.startTime = DateTime.MinValue + (DateTime.Now - info.startTime); // apply of pause delta time
         }
      }

      public void Pause(object subscriber) {
         SubscribeInfo info = _subscribers[subscriber];
         if (info == null)
            return;
         info.active = false;
         info.startTime = DateTime.MinValue + (DateTime.Now - info.startTime); // set of pause delta time
      }

      public void Unsubscribe(object subscriber) {
         _subscribers.Remove(subscriber);
      }

      public void Dispose() {
         _unsubscribe.Dispose();
         _subscribers.Clear();
      }

   }

}

File: src/src_platform_jfx/fmg/jfx/img/Animator.java
package fmg.jfx.draw.img;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

import javafx.animation.AnimationTimer;

import fmg.core.img.IAnimator;

public class Animator implements IAnimator, AutoCloseable {

   private static class SubscribeInfo {
      public boolean active = true;    // enabled?
      public long    startTime = new Date().getTime(); // start time of subscribe
      public Consumer<Long /* time from the beginning of the subscription */> callback;
   }
   private final AnimationTimer _timer;
   private final Map<Object /* subscriber */, SubscribeInfo> _subscribers;

   private static Animator _singleton;
   public static Animator getSingleton() { // not synchronized. since should work only in the thread of the UI.
      if (_singleton == null)
         _singleton = new Animator();
      return _singleton;
   }

   private Animator() {
      _subscribers = new HashMap<>();
      _timer = new AnimationTimer() {

          @Override
          public void handle(long now) {
              long currentTime = new Date().getTime();
              _subscribers.forEach((k,v) -> {
                 if (v.active)
                    v.callback.accept(currentTime - v.startTime);
              });
          }

      };
      _timer.start();
   }

   @Override
   public void subscribe(Object subscriber, Consumer<Long /* time from start subscribe */> subscriberCallbackMethod) {
      SubscribeInfo info = _subscribers.get(subscriber);
      if (info == null) {
         info = new SubscribeInfo();
         info.callback = subscriberCallbackMethod;
         _subscribers.put(subscriber, info);
      } else {
         info.active = true;
         info.startTime = new Date().getTime() - info.startTime; // apply of pause delta time
      }
   }

   @Override
   public void pause(Object subscriber) {
      SubscribeInfo info = _subscribers.get(subscriber);
      if (info == null)
         return;
      info.active = false;
      info.startTime = new Date().getTime() - info.startTime; // set of pause delta time
   }

   @Override
   public void unsubscribe(Object subscriber) {
      _subscribers.remove(subscriber);
   }

   @Override
   public void close() {
      _timer.stop();
      _subscribers.clear();
   }

}
